### ISSUE TITLE
bigip_virtual_server is very slow after migrating from Ansible 2.7 to Ansible 2.8 or 2.9

### ISSUE THREAD

**Issue by slavkv** [F5Networks/f5-ansible#1610]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

bigip_virtual_server

##### ANSIBLE VERSION

```
ansible 2.8.7
  config file = /Users/user1/git/gtld-load-balancers-automation/ansible.cfg
  configured module search path = [u'/Users/user1/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/user1/git/ansible-2.8-virt/lib/python2.7/site-packages/ansible
  executable location = /Users/user1/git/ansible-2.8-virt/bin/ansible
  python version = 2.7.16 (default, Sep  2 2019, 12:00:52) [GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.11.45.5)]
```

##### PYTHON VERSION

```
Python 2.7.16
Python 3.7
```

##### BIGIP VERSION

```
Main Package
  Product     BIG-IP
  Version     14.1.2.3
  Build       0.0.5
  Edition     Point Release 3
  Date        Tue Dec 17 19:48:55 PST 2019
```

##### CONFIGURATION

$ cat ansible.cfg
[defaults]
inventory = ./inventories/staging
host_key_checking = False
ask_pass=False
roles_path = ./roles:/usr/local/etc/ansible/roles/
retry_files_enabled = False

##### OS / ENVIRONMENT

N/A

##### SUMMARY

I use Ansible module `bigip_virtual_server` to configure virtual servers.
It was working good, relative ok in terms of speed.

After I started using Ansible 2.8 or Ansible 2.9 all modules work the same except of `bigip_virtual_server`, it takes ages now to execute. When I have a role with many SSL profiles on a Virtual Server, it never completes.

I tried different versions of Python (ver 2 and ver 3) and Ansible 2.8 and 2.9. The same result, although 2.9 takes a little bit faster.

I took traffic dumps of those communications to compare what's going on on network level.
I tested `bigip_virtual_server` and `bigip_pool` for comparison.

Below I provided sizes of tcpdump files (HTTPS traffic between Ansible and F5, port 443 only; there was no other communication on this port between these hosts during tests).
```
Ansible 2.7:
bigip_pool: 24KB size, 2 TCP sessions, time to execute is 0m3.635s
bigip_virtual_server: 105KB size, 15 TCP sessions, time to execute is 0m5.800s
```

```
Ansible 2.8:
bigip_pool: 23KB size, 4 TCP sessions, time to execute is 0m3.611s
bigip_virtual_server: 63MB size !!, 145 TCP sessions, time to execute is 1m13.276s
```

```
Ansible 2.9:
bigip_pool: 23KB size, 4 TCP sessions, time to execute is 0m3.346s
bigip_virtual_server: 39MB size !!, ?? TCP sessions, time to execute is 0m45.710s
```

As you can see, there is no difference for module `bigip_pool` while there is a huge difference for `bigip_virtual_server`, it takes **more than 40 seconds** to update VS and **generate 40+ Megabytes of traffic** for it, which is not normal.

##### STEPS TO REPRODUCE

Even you see `loop` operator below please keep in mind that I tested on a list with a single item.
```
- name: Create pools
  bigip_pool:
    provider: "{{ f5_cli }}"
    state: present
    description: "{{ pool.description | default('') }}"
    name: "{{ pool.name }}"
    partition: Common
    lb_method: round-robin
    monitor_type: single
    monitors:
      - "{{ pool.monitor }}"
  delegate_to: localhost
  notify: Save F5 configuration
  loop: "{{ ltm.pools }}"
  loop_control:
    loop_var: pool
    label: " Create pool {{ pool.name }} "

- name: Add virtual servers (with irules, no policies)
  bigip_virtual_server:
    provider: "{{ f5_cli }}"
    state: "{{ vs.state | default('present') }}"
    partition: Common
    name: "{{ vs.name }}"
    type: "{{ vs.type | default('standard') }}"
    ip_protocol: "{{ vs.ip_protocol | default('tcp')}}"
    destination: "{{ vs.destination }}"
    port: "{{ vs.port }}"
    pool: "{{ vs.pool | default ('') }}"
    description: "{{ vs.description | default('') }}"
    profiles: "{{ vs.profiles }}"
    default_persistence_profile: "{{ vs.persist | default('') }}"
    irules: "{{ vs.irules | default(['']) }}"
    snat: "{{ vs.snat }}"
    enabled_vlans: "{{ vs.enabled_vlans | default('all')}}"
  delegate_to: localhost
  notify: Save F5 configuration
  loop: "{{ ltm.virtuals }}"
  loop_control:
    loop_var: vs
    label: " Create virtual server {{ vs.name }} "
```

##### EXPECTED RESULTS

A `bigip_virtual_server` (in Ansible 2.8 and 2.9) can configure Virtual Server with approximately the same time as Ansible 2.7, and it must not generate so much network traffic for a single Virtual Server.

##### ACTUAL RESULTS

```
Ansible 2.7:
bigip_pool: 24KB size, 2 TCP sessions, time to execute is 0m3.635s
bigip_virtual_server: 105KB size, 15 TCP sessions, time to execute is 0m5.800s
```

```
Ansible 2.8:
bigip_pool: 23KB size, 4 TCP sessions, time to execute is 0m3.611s
bigip_virtual_server: 63MB size !!, 145 TCP sessions, time to execute is 1m13.276s
```

```
Ansible 2.9:
bigip_pool: 23KB size, 4 TCP sessions, time to execute is 0m3.346s
bigip_virtual_server: 39MB size !!, ?? TCP sessions, time to execute is 0m45.710s


When I specify lots of client-ssl profiles for a Virtual Server, it never finishes and times out.
In tcpdump I see that traffic is going on all this time.

<!--- Paste verbatim command output between quotes below -->
```

```

**Comment by focrensh** [comment]
As new features and parameters get added to modules they will increase in execution time as more API calls must be made to ensure idempotency and then configure the object.

That being said, i ran your exact example above and it is taking less than 15s to either create or verify the existing VS.

Will you provide an example with the number of client-ssl profiles to takes the have the module not finish?

**Comment by slavkv** [comment]
@focrensh sure, I will provide full code I do. Give me some time :)

**Comment by slavkv** [comment]
@focrensh
I obfuscated code and upload it https://github.com/slavkv/debug-ansible-f5
Please test.

I generated 100 SSL keys and certs and configuring VS with 100 client-ssl profiles.

It takes 14 minutes to do this one task. It is very slow and something is wrong, I am not sure where. Any help will be much appreciated.

I run this repository and provided my numbers in README file.

Thanks!

**Comment by slavkv** [comment]
I think I found why this takes so much time.

When `ModuleParameters.profiles` property is called it iterates though all passed profiles to be configured on a Virtual Server.

For each of the profiles it calls two methods(?):
1. `self._handle_profile_context(tmp)`
2. `self._handle_ssl_profile_nuances(tmp)`

The problem is with the second one.

```   
def _handle_ssl_profile_nuances(self, profile):
        if profile['name'] == 'serverssl' or self._is_server_ssl_profile(profile):
            if profile['context'] != 'serverside':
                profile['context'] = 'serverside'
        if profile['name'] == 'clientssl' or self._is_client_ssl_profile(profile):
            if profile['context'] != 'clientside':
                profile['context'] = 'clientside'
        return
```

For each client-ssl profile it calls `self._is_client_ssl_profile(profile)` to check if it exists on F5.
And that is why it takes so much time.

``` 
def _is_client_ssl_profile(self, profile):
        if profile['name'] in self._read_current_clientssl_profiles_from_device():
            return True
        return False
```

```
    def _read_current_clientssl_profiles_from_device(self):
        uri = "https://{0}:{1}/mgmt/tm/ltm/profile/client-ssl/".format(
            self.client.provider['server'],
            self.client.provider['server_port'],
        )
        resp = self.client.api.get(uri)
        try:
            response = resp.json()
        except ValueError as ex:
            raise F5ModuleError(str(ex))

        if 'code' in response and response['code'] == 400:
            if 'message' in response:
                raise F5ModuleError(response['message'])
            else:
                raise F5ModuleError(resp.content)
        result = [x['name'] for x in response['items']]
        return result
```

It calls F5 **every** time for each client-ssl profile. Which adds about 1 second.
And `_read_current_clientssl_profiles_from_device` gets **all client-ssl profiles** from F5.

Plus, `ModuleParameters.profiles` is called **16 times** !!

So, if we have 200 client-ssl profiles to be configured on VS:
1. Around **200 seconds** to check each client-ssl profile exists on F5. It equals to **3-4 mins**. This explains why there is SO much traffic is in place between Ansible and F5.
2. `ModuleParameters.profiles` is called 16 times, **16 * 4 min = 48 mins** just to check that all client-ssl profiles exist on F5 and configure VS. It is terrible.


If I remove `or self._is_client_ssl_profile(profile)` part from `_handle_ssl_profile_nuances(self, profile):` it takes about **7-8 seconds** to configure VS by Ansible, and `ModuleParameters.profiles` is still called 16 times.
Ansible 2.7 does not have this check and that's why it works fast.

**Comment by slavkv** [comment]
Do not get me wrong, I understand the point of checking if provided client-ssl (or server-ssl) exists on F5. It makes sense.

I am just saying it should not be like that and it is a bug which heavily effects on Ansible performance.


The logic in the module should be the following:
1. `bigip_virtual_server` detects if profiles can be configured (based on VS type)
2. `bigip_virtual_server` gets a list of all profiles on F5 (a few API calls only, not for each requested SSL profile).
3. `bigip_virtual_server` compares this list with the list of requested profiles for VS

Instead of calling F5 to get a list of all client-ssl certs for every provided profile, a module should get a list of all client-ssl once, and then it can compare it to the list of profiles we want to configure.

**Comment by slavkv** [comment]
Another option is to remove this `or self._is_client_ssl_profile(profile)` part.

What value does it provide?
1. When it is used it slows down performance drastically and does not finish (when need to configure a lot of profiles). **In this case I do not get a clear error why it took so much time.**
2. When it is not used and a requested clien-ssl profile does not exist, `bigip_virtual_server` fails immediately and the trace clearly says that a specific client-ssl profile was not found.

If people choose between 30 mins timeout and 8 sec failure with the error they understand and can fix, they will choose option 2, aren't they?

**Comment by slavkv** [comment]
And the last workaround is to introduce the switch/parameter into the module which will indicate if you want `bigip_virtual_server` to check profiles exist on F5 or not. You can enable it by default but it should be possible to turn this off.

It is not a bug fix but a solution which will help in these scenarios.

Please let me know what you think guys and help with the fix.

**Comment by focrensh** [comment]
Thank you for the investigation. We will investigate performance options here.

Added to our backlog under FMFA-423

**Comment by slavkv** [comment]
@focrensh thank you. What is the procedure for backlog issues? What's the change a fix can be introduced in several months?

**Comment by focrensh** [comment]
We are discussing options (like the flag) internally. I would say the odds of a couple months into a collection are very high. We will update this issue once the updates are made to the daily collection build which then can be used for testing.

**Comment by slavkv** [comment]
@focrensh thank you for your answers and all work you do guys. cheers.

**Comment by diLLec** [comment]
@slavkv With your investigation in mind (thanks alot btw!), I've added a simple caching step into both functions. This may be better solved by using the ansible caching mechanism (see https://docs.ansible.com/ansible/latest/plugins/cache.html#using-cache-plugins). For me the following code reduces the execution time of the virtual server module from 45-60 seconds to 5-10 while having the same result

```
    read_current_clientssl_profiles_from_device_cache_filled = False
    read_current_clientssl_profiles_from_device_cache = []

    def _read_current_clientssl_profiles_from_device(self):
        if self.read_current_clientssl_profiles_from_device_cache_filled:
            return self.read_current_clientssl_profiles_from_device_cache

        uri = "https://{0}:{1}/mgmt/tm/ltm/profile/client-ssl/".format(
            self.client.provider['server'],
            self.client.provider['server_port'],
        )
        resp = logged_api_get(self,uri)
        try:
            response = resp.json()
        except ValueError as ex:
            raise F5ModuleError(str(ex))

        if 'code' in response and response['code'] == 400:
            if 'message' in response:
                raise F5ModuleError(response['message'])
            else:
                raise F5ModuleError(resp.content)
        self.read_current_clientssl_profiles_from_device_cache = [x['name'] for x in response['items']]
        self.read_current_clientssl_profiles_from_device_cache_filled = True
        return self.read_current_clientssl_profiles_from_device_cache

    read_current_serverssl_profiles_from_device_cache_filled = False
    read_current_serverssl_profiles_from_device = []

    def _read_current_serverssl_profiles_from_device(self):
        if self.read_current_serverssl_profiles_from_device_cache_filled:
            return self.read_current_serverssl_profiles_from_device_cache

        uri = "https://{0}:{1}/mgmt/tm/ltm/profile/server-ssl/".format(
            self.client.provider['server'],
            self.client.provider['server_port'],
        )
        resp = logged_api_get(self,uri)
        try:
            response = resp.json()
        except ValueError as ex:
            raise F5ModuleError(str(ex))

        if 'code' in response and response['code'] == 400:
            if 'message' in response:
                raise F5ModuleError(response['message'])
            else:
                raise F5ModuleError(resp.content)
        self.read_current_serverssl_profiles_from_device_cache = [x['name'] for x in response['items']]
        self.read_current_serverssl_profiles_from_device_cache_filled = True
        return self.read_current_serverssl_profiles_from_device_cache

```

**Comment by slavkv** [comment]
@diLLec thank you. Can we expect to have this fix in next Ansible 2.8.* or 2.9.* maintenance release?

**Comment by wojtek0806** [comment]
hello, there will be no fix in the next release, modules are removed from ansible and moved into collections, once we asses our options fix will be provided, no ETA
